// @tanstack/start-plugin-core/dist/esm/vite/start-manifest-plugin/plugin.js
import { ENTRY_POINTS, START_ENVIRONMENT_NAMES } from "../../constants.js";
import { createVirtualModule } from "../createVirtualModule.js";
import { buildStartManifest, serializeStartManifest } from "../../start-manifest-plugin/manifestBuilder.js";
import { VIRTUAL_MODULES } from "@tanstack/start-server-core";
import { joinURL } from "ufo";
//#region src/vite/start-manifest-plugin/plugin.ts
function startManifestPlugin(opts) {
	return createVirtualModule({
		name: "tanstack-start:start-manifest-plugin",
		moduleId: VIRTUAL_MODULES.startManifest,
		enforce: "pre",
		load() {
			const { resolvedStartConfig } = opts.getConfig();
			if (this.environment.name !== START_ENVIRONMENT_NAMES.server) return "export default {}";
			if (this.environment.config.command === "serve") return `export const tsrStartManifest = () => ({
            routes: {},
            clientEntry: '${joinURL(resolvedStartConfig.basePaths.publicBase, "@id", ENTRY_POINTS.client)}',
          })`;
			const routeTreeRoutes = globalThis.TSS_ROUTES_MANIFEST;
			const clientBuild = opts.getClientBuild();
			if (!clientBuild) return `export const tsrStartManifest = () => ({
              routes: {},
              clientEntry: '${joinURL(resolvedStartConfig.basePaths.publicBase, "@id", ENTRY_POINTS.client)}',
              })`;
			return `export const tsrStartManifest = () => (${serializeStartManifest(buildStartManifest({
				clientBuild,
				routeTreeRoutes,
				basePath: resolvedStartConfig.basePaths.publicBase
			}))})`;
		}
	});
}
//#endregion
export { startManifestPlugin };
